File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = -1
    
    for dungeon in permutations(dungeons, len(dungeons)):
        tmp = k
        cnt = 0
        
        for a, b in dungeon:
            if tmp >= a:
                tmp -= b
                cnt += 1
        
        answer = max(cnt, answer)
    
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int answer = -1;
    static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(dungeons, k, 0);
        return answer;
    }
    
    public static void dfs(int[][] dungeons, int k, int depth) {
        for(int i=0; i<dungeons.length; i++) {
            if(!visited[i] && dungeons[i][0] <= k) {
                visited[i] = true;
                dfs(dungeons, k - dungeons[i][1], depth+1);
                visited[i] = false;
            }
        }
        
        answer = Math.max(answer, depth);
    }
}